pkg/dynamicinstrumentation/uploader/writer.go

```
@@ -0,0 +1,72 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf && arm64
+
+package uploader
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/diagnostics"
+	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/ditypes"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type WriterSerializer[T any] struct {
+	output io.Writer
+	mu     sync.Mutex
+}
+
+func NewWriterLogSerializer(writer io.Writer) (*WriterSerializer[ditypes.SnapshotUpload], error) {
+	if writer == nil {
+		return nil, errors.New("nil writer for creating log serializer")
+	}
+	return NewWriterSerializer[ditypes.SnapshotUpload](writer)
+}
+
+func NewWriterDiagnosticSerializer(dm *diagnostics.DiagnosticManager, writer io.Writer) (*WriterSerializer[ditypes.DiagnosticUpload], error) {
+	if writer == nil {
+		return nil, errors.New("nil writer for creating diagnostic serializer")
+	}
+	ds, err := NewWriterSerializer[ditypes.DiagnosticUpload](writer)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		for diagnostic := range dm.Updates {
+			ds.Enqueue(diagnostic)
+		}
+	}()
+	return ds, nil
+}
+
+func NewWriterSerializer[T any](writer io.Writer) (*WriterSerializer[T], error) {
+	if writer == nil {
+		return nil, errors.New("nil writer for creating serializer")
+	}
+	return &WriterSerializer[T]{
+		output: writer,
+	}, nil
+}
+
+func (s *WriterSerializer[T]) Enqueue(item *T) bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	bs, err := json.Marshal(item)
+	if err != nil {
+		log.Error("Failed to marshal item", item)
```

IMHO it would be better to return errors, rather than unconditionally logging.